File: android/app/src/main/kotlin/io/github/friesi23/mhabit/collation/CollationPlugin.kt
package io.github.friesi23.mhabit.collation

import android.icu.text.Collator
import android.icu.util.ULocale

class CollationPlugin : CollationHostApi {
    override fun sortStrings(request: CollationRequest): List<String> {
        val uLocale = request.locale
            ?.takeIf { it.isNotBlank() }
            ?.let { ULocale.forLanguageTag(it.replace("_", "-")) }
            ?.setKeywordValue("colNumeric", "yes")
            ?: ULocale.getDefault().setKeywordValue("colNumeric", "yes")

        val collator = Collator.getInstance(uLocale).apply {
            strength = Collator.TERTIARY
            decomposition = Collator.CANONICAL_DECOMPOSITION
        }

        return sortByCollationKey(request.items) { collator.getCollationKey(it) }
    }
}

File: darwin/CollationPlugin/Sources/CollationPlugin/CollationPlugin.swift
import Foundation

class CollationPlugin: CollationHostApi {
    func sortStrings(request: CollationRequest) throws -> [String] {
        let entries = request.items.map { item in
            (id: item.id, value: item.value)
        }

        let sorted = entries.sorted { a, b in
            a.value.localizedStandardCompare(b.value) == .orderedAscending
        }

        return sorted.map(\.id)
    }
}
